src/notebooksyncagent.cpp

```diff
@@ -264,11 +267,70 @@ void NotebookSyncAgent::fetchRemoteChanges(const QDateTime &fromDateTime, const
     report->getAllETags(mServerPath, fromDateTime, toDateTime);
 }
 
+// A given incidence has been added or modified locally.
+// To upsync the change, we need to construct the .ics data to upload to server.
+// Since the incidence may be an occurrence or recurring series incidence,
+// we cannot simply convert the incidence to iCal data, but instead we have to
+// upsync an .ics containing the whole recurring series.
+QString NotebookSyncAgent::constructLocalChangeIcs(KCalCore::Incidence::Ptr updatedIncidence)
+{
+    // create an in-memory calendar
+    // add to it the required incidences (ie, check if has recurrenceId -> load parent and all instances; etc)
+    // for each of those, we need to do the IncidenceToExport() modifications first
+    // then, export from that calendar to .ics file.
+    KCalCore::MemoryCalendar::Ptr memoryCalendar(new KCalCore::MemoryCalendar(KDateTime::UTC));
+    if (updatedIncidence->hasRecurrenceId() || updatedIncidence->recurs()) {
```

Comment:
I think the recurrence ids should be updated.
Eg: The main event starts at 6pm every day, and the third occurrence is a persistent/exception occurrence which starts instead at 8pm.  The user then modifies the main event to start at 7pm - if we don't update both the ExDate in the main event AND the RecurrenceId in the persistent exception occurrence, then our calendar has an invalid event (the exception is now "floating" - it's recurrenceId doesn't actually refer to any valid occurrence date of the original main event).